data_model/primitives/src/conststr.rs

```diff
@@ -0,0 +1,650 @@
+        deserializer.deserialize_str(ConstStringVisitor)
+    }
+}
+
+struct ConstStringVisitor;
+
+impl<'de> Visitor<'de> for ConstStringVisitor {
+    type Value = ConstString;
+
+    fn expecting(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.write_str("a string")
+    }
+
+    fn visit_str<E: Error>(self, v: &str) -> Result<Self::Value, E> {
+        Ok(v.into())
+    }
+
+    fn visit_string<E: Error>(self, v: String) -> Result<Self::Value, E> {
+        Ok(v.into())
+    }
+}
+
+impl WrapperTypeEncode for ConstString {}
+
+impl WrapperTypeDecode for ConstString {
+    type Wrapped = String;
+}
+
+impl IntoSchema for ConstString {
+    fn type_name() -> String {
+        String::type_name()
+    }
+    fn schema(map: &mut MetaMap) {
+        String::schema(map);
+    }
+}
+
+#[derive(DebugCustom)]
+#[debug(fmt = "{:?}", "&**self")]
+#[repr(C)]
+struct BoxedString {
+    #[cfg(target_endian = "little")]
+    ptr: *mut u8,
+    len: usize,
+    #[cfg(target_endian = "big")]
+    ptr: *mut u8,
+}
+
+impl BoxedString {
+    #[inline]
+    const fn len(&self) -> usize {
+        self.len
+    }
+
+    #[allow(unsafe_code)]
+    #[inline]
+    fn as_bytes(&self) -> &[u8] {
+        // SAFETY: created from `Box<[u8]>`.
+        unsafe { from_raw_parts(self.ptr, self.len) }
+    }
+}
+
+impl AsRef<str> for BoxedString {
+    #[allow(unsafe_code)]
+    #[inline]
+    fn as_ref(&self) -> &str {
+        // SAFETY: created from valid utf-8
+        unsafe { from_utf8_unchecked(self.as_bytes()) }
+    }
+}
+
+impl Deref for BoxedString {
+    type Target = str;
+
+    #[inline]
+    fn deref(&self) -> &Self::Target {
+        self.as_ref()
+    }
+}
+
+impl Clone for BoxedString {
+    /// Properly clone `BoxedString` into new allocation.
+    fn clone(&self) -> Self {
+        let ptr = Box::into_raw(self.as_bytes().to_owned().into_boxed_slice()).cast::<u8>();
+        let len = self.len;
+        Self { ptr, len }
+    }
+}
+
+impl From<&str> for BoxedString {
+    #[allow(unsafe_code)]
+    #[inline]
+    fn from(value: &str) -> Self {
+        let len = value.len();
+        let boxed = value.as_bytes().to_vec().into_boxed_slice();
+        let ptr = Box::into_raw(boxed).cast::<u8>();
+        Self { ptr, len }
+    }
+}
+
+impl From<String> for BoxedString {
+    #[inline]
+    fn from(value: String) -> Self {
+        let len = value.len();
+        let boxed = value.into_bytes().into_boxed_slice();
+        let ptr = Box::into_raw(boxed).cast::<u8>();
+        Self { ptr, len }
+    }
+}
+
+impl Drop for BoxedString {
+    #[allow(unsafe_code)]
+    fn drop(&mut self) {
+        // SAFETY: created from `Box<[u8]>`.
+        unsafe {
+            Box::<[_]>::from_raw(from_raw_parts_mut(self.ptr, self.len));
+        }
+    }
+}
+
+/// `BoxedString` is `Send` because the data they
+/// reference is unaliased. Aliasing invariant is enforced by
+/// creation of `BoxedString`.
+#[allow(unsafe_code)]
+unsafe impl Send for BoxedString {}
```

Comment:
If [`core::ptr::Unique`](https://docs.rs/rust-libcore/latest/core/ptr/struct.Unique.html) (which `Box` uses under the hood) was stable feature `BoxedString` could use them and will be automatically `Send` and `Sync`, but `Sync` and `Send` is not inplemented for pointers (and also `NonNull`), so we must implement them manually.
We can do that because every `BoxedString` instance has own unique pointer (there is no two `BoxedString`s pointing) to the same memory. 

Or what you mean by? 
> does this mean ConstString is Send and Sync transitively?

Reply:
Are you going to remove the implementations once they are automatic, based off of the feature that you linked?

Reply:
my question was is `ConstString` both `Send` and `Sync`? Not about `BoxedString`

Reply:
I think `ConsString` is both `Send` and `Sync`. Verify and resolve this

Reply:
Both `BoxedString` and `InlinedString` are `Send` and `Sync` so yes `ConstString` is also `Send` and `Sync` transitively.